extensions/infinispan-cache/runtime/src/main/java/io/quarkus/cache/infinispan/runtime/InfinispanCacheImpl.java

```diff
@@ -0,0 +1,211 @@
+                            resultAsync.completeExceptionally((Throwable) ex2);
+                        } else if (existing == null) {
+                            resultAsync.complete(newValue);
+                        } else {
+                            resultAsync.complete(decodeNull(existing));
+                        }
+                        computationResults.remove(key);
+                    });
+            return resultAsync;
+        }));
+    }
+
+    @Override
+    public <K, V> Uni<V> getAsync(K key, Function<K, Uni<V>> valueLoader) {
+        return Uni.createFrom().completionStage(handleAndCompose(remoteCache.getAsync(key), (v1, ex1) -> {
+            if (ex1 != null) {
+                return CompletableFuture.failedFuture(ex1);
+            }
+
+            if (v1 != null) {
+                return CompletableFuture.completedFuture(decodeNull(v1));
+            }
+
+            CompletableFuture<V> resultAsync = new CompletableFuture<>();
+            CompletableFuture<V> computedValue = computationResults.putIfAbsent(key, resultAsync);
+            if (computedValue != null) {
+                return computedValue;
+            }
+            valueLoader.apply(key).convert().toCompletionStage()
+                    .whenComplete((newValue, ex2) -> {
+                        if (ex2 != null) {
+                            resultAsync.completeExceptionally(ex2);
+                            computationResults.remove(key);
+                        } else {
+                            remoteCache.putIfAbsentAsync(key, encodeNull(newValue), lifespan, TimeUnit.MILLISECONDS, maxIdle,
+                                    TimeUnit.MILLISECONDS).whenComplete((existing, ex3) -> {
+                                        if (ex3 != null) {
+                                            resultAsync.completeExceptionally((Throwable) ex3);
+                                        } else if (existing == null) {
+                                            resultAsync.complete(newValue);
+                                        } else {
+                                            resultAsync.complete(decodeNull(existing));
+                                        }
+                                        computationResults.remove(key);
+                                    });
+                        }
+                    });
+            return resultAsync;
+        }));
+    }
+
+    @Override
+    public Uni<Void> invalidate(Object key) {
+        return Uni.createFrom().completionStage(remoteCache.removeAsync(key));
+    }
+
+    @Override
+    public Uni<Void> invalidateAll() {
+        return Uni.createFrom().completionStage(remoteCache.clearAsync());
+    }
+
+    @Override
+    public Uni<Void> invalidateIf(Predicate<Object> predicate) {
+        return Uni.createFrom().item(new Supplier<Void>() {
+            @Override
+            public Void get() {
+                remoteCache.keySet().removeIf(predicate);
```

Comment:
This is blocking, which I am guessing will be an issue :(

Reply:
Is there a way to mark this as required to be offloaded to a blocking thread?

Reply:
@cescoffier @geoand

Reply:
Is there to way to have `removeCache.keySetAsync()`?

Reply:
Unrelated to this PR, I'll look more close, the RedisImpl implementation uses "Keys" 
https://redis.io/commands/keys/

Reply:
Right, but in the RedisClient we use, all operations are async.

This (and any other cache implementation) will also have to ensure that all Network IO is non-blocking.

Reply:
I know, I'm just pointing we are using a command that is not recommended for this use case. The command is executed in the server after all

Reply:
for what is the blocking of the client, as I mentioned above, I believed this method was not blocking underneath. We will do it in another way :)
In any case the this invalidation with predicate, does not look as something that I would use to be honest, but that's my call

Reply:
> In any case this invalidation with predicate, does not look as something that I would use to be honest, but that's my call

Right, I totally understand that, But for better or worse it's part of the API

Reply:
I changed the method using the executor

Reply:
I am not sure @cescoffier will like that 🙂

Reply:
I would never recommend to use this method xD